File: .backup/implementations/rocket.ts
import { log } from "../common/adds";
import { DatableI, GalaxyTouchingObjectsI, RocketFireI, RocketTypeI, RocketI, DataCalcI } from "../common/declarations";
import { inRange, Vector } from "../common/math";
import { rocketTypes } from "../common/rocketTypes";
import { Galaxy } from "./galaxy";
import { MovingObject } from "./movingObject";
import { RocketFire } from "./rocketFire";
import { User } from "./user";

export class Rocket extends MovingObject implements RocketI, DatableI<RocketI> {
    img = '' // never used!
    touchingObjects: GalaxyTouchingObjectsI = {
        asteroids: []
    }

    fires: RocketFire[] = []
    
    myUser: () => User

    rocketType: RocketTypeI = rocketTypes('standart-rocket')
    rocketTypeId: rocketTypes = this.rocketType.id

    constructor(myUser: () => User, myGalaxy: () => Galaxy, id?: number) {
        super(myGalaxy, id)
        if (id) this.id = id
        this.myUser = myUser
        this.setRocketType('standart-rocket')
    }

    setRocketType(t: rocketTypes) {
        this.rocketType = rocketTypes(t)
        this.rocketTypeId = t
        this.fires = this.rocketType.fires.map(e => new RocketFire(e))
        this.img = this.rocketType.img
        this.geo.width = this.rocketType.width
        this.geo.height = this.rocketType.height
    }

    calc(s: number) {
        const ar = this.myUser().checkKey('ArrowRight')
        const al = this.myUser().checkKey('ArrowLeft')
        const au = this.myUser().checkKey('ArrowUp')

        if (au) this.fires.forEach(f => f.calc(au, this.geo))

        if (ar === true || al === true)
            this.geo.angle = this.geo.angle + (ar ? 1 : -1)*this.rocketType.turningSpeed*s

        if (au === true) {
            this.movingVector = this.movingVector.plus(
                Vector.fromAL(
                    this.geo.angle, 
                    this.rocketType.acceleratingSpeed
                ).mul(s)
            )
            console.log('accelerating')
        }

        super.calc(s)
    }

    data() {
        return {
            ...super.data(),
            touchingObjects: this.touchingObjects,
            rocketTypeId: this.rocketTypeId
        }
    }

    static zoomSpeed = 0.05
}

const zoomSpeed = 0.05

interface RocketSI extends RocketI {
    rocketType: RocketTypeI
}

function Rocket(r: RocketSI) {
    return {
        setRocketType(t: rocketTypes) {
            r.rocketType = rocketTypes(t)
            r.rocketTypeId = t
            r.fires = r.rocketType.fires.map(e => new RocketFire(e))
            r.img = r.rocketType.img
            r.geo.width = r.rocketType.width
            r.geo.height = r.rocketType.height
        },
        stringify() {
            let rr = { ...r }
            delete rr.rocketType
            return rr
        }
    }
}

function RocketFire(params:type) {
    
}

File: src/main/kotlin/server/game/objects/Rocket.kt
package server.game.objects

import server.adds.math.CrazyTransform
import server.adds.math.geom.GeoI
import server.adds.math.RocketMath.inRange
import server.adds.math.CrazyVector
import server.adds.math.geom.shapes.CrazyPolygon
import server.adds.math.geom.shapes.CrazyShape
import server.adds.math.vec
import server.data_containers.KeyboardI
import server.data_containers.UserViewI
import server.data_containers.*

class Rocket(
    pos: CrazyVector,
    val userProps: UserPropsI,
    id: String,
) : GeoObject(pos, CrazyVector.zero(), 0.0, CrazyVector.zero(), id, GameObjectType.ROCKET) {
    private var keyboard = KeyboardI()
    private lateinit var rocketType: RocketType
    private var fires = listOf<RocketFire>()

    var eye: CrazyVector
    var zoom = 1.0
    var zoomTarget = 1.0

    init {
        setRocketType(RocketType.DEFAULT)
        eye = pos
    }

    fun userView() = UserViewI(eye, zoom)

    private fun setRocketType(t: RocketType) {
        rocketType = t
        fires = rocketType.fires.map { RocketFire(it) }
        /*width = rocketType.width
        height = rocketType.height*/
        zoomTarget = rocketType.defaultZoom
    }

    fun setKeyboard(k: KeyboardI) {
        keyboard = k
    }

    override fun collider() = CrazyPolygon(
        arrayOf(
            vec(-0.5, 0.5),
            vec(0.0, -0.5),
            vec(0.5, 0.5)
        )
    ).convert { (it * size) rotate ang }

    override suspend fun calc(s: Double) {
        val arrowUp = keyboard.key("ArrowUp")
        val arrowRight = keyboard.key("ArrowRight")
        val arrowLeft = keyboard.key("ArrowLeft")

        /*if (arrowUp) this.fires.forEach {
            it.owner = this.getGeo().copy()
            it.calc(s)
        }*/

        if (arrowRight || arrowLeft)
            ang += (if (arrowRight) 1 else -1) * rocketType.turningSpeed * s

        if (arrowUp) velocity += vec(ang, rocketType.acceleratingSpeed, true) * s

        eye += (pos - eye) * rocketType.eyeLazy * s

        /*log("$velocity")
        log("$pos")
        log("$eye")
        log("")*/

        super.calc(s)
    }

    override fun data() = RocketOI(
        userProps = userProps,
        geo = getGeo(),
        id = id,
        view = UserViewI(eye, zoom),
        style = RocketStyleI(
            img = rocketType.img,
            fires = fires.map { it.data() }.toTypedArray()
        )
    )

    class RocketFire(
        private val settings: RocketFireSettingsI
    ) : GameClassI {
        private var on = false
        private var fireShown = 0.0
        private var fireTarget = 0.0

        private var geo = GeoI()
        var owner = GeoI()

        override suspend fun calc(s: Double) {
            val f = this.fireShown / this.fireTarget
            geo = geo.copy(
                width = this.settings.startWidth + (this.settings.plusWidth)*f,
                height = this.settings.startHeight + (this.settings.plusWidth)*f
            )

            if (inRange(fireShown, fireTarget, settings.fireSpeed)) fireShown = fireTarget
            else fireShown += settings.fireSpeed * (if (fireShown > fireTarget) -1 else 1)

            if (fireShown != 0.0) {
                val it = ((1 - this.fireShown + 0.5)* 0.3 * 0.4)
                geo = geo.copy(ang = Math.random()*it - it/2.0)
            }

            geo = geo.copy(pos = vec(
                settings.dx,
                settings.dy + owner.height/2
                        - geo.height/2
                        + geo.height
                        * fireShown * 0.9 - 0.5
            )
            )

            if (on && fireShown != 1.0) fireShown = 1.0
            else if (!on && fireShown != 0.0) fireShown = 0.0
        }

        override fun data() = RocketFireI(on, geo, settings.img, "never-used")
    }
}